src/v/storage/disk_log_impl.cc

```diff
@@ -256,10 +259,104 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
         return storage::internal::self_compact_segment(seg, cfg, _probe)
           .finally([seg] { seg->mark_as_finished_self_compaction(); });
     }
-    // all segments are self-compacted
-    // do cross segment compaction
+
+    if (auto range = find_compaction_range(); range) {
+        return compact_adjacent_segments(std::move(*range), cfg);
+    }
+
     return ss::now();
 }
+
+std::optional<std::pair<segment_set::iterator, segment_set::iterator>>
+disk_log_impl::find_compaction_range() {
+    /*
+     * adjacent segment compaction.
+     *
+     * the strategy is to choose a pair of adjacent segments and first combine
+     * them into a single segment that replaces the pair, and then perform
+     * self-compaction on the replacement segment.
+     */
+    if (_segs.size() < 2) {
+        return std::nullopt;
+    }
+
+    // sliding window over segments. currently restricted to two segments
+    auto range = std::make_pair(_segs.begin(), std::next(_segs.begin(), 2));
+
+    do {
+        // the simple compaction process in use right now builds a concatenation
+        // of segments so we avoid processing a group that is too large.
+        auto size = std::accumulate(
+          range.first,
+          range.second,
+          size_t(0),
+          [](size_t acc, ss::lw_shared_ptr<segment>& seg) {
+              return acc + seg->size_bytes();
+          });
+        if (size < _manager.config().max_compacted_segment_size) {
+            break;
+        }
+
+        // advance the window
+        ++range.first;
+        ++range.second;
+    } while (range.second != _segs.end());
+
+    // no candidate pair found
+    if (range.second == _segs.end()) {
+        return std::nullopt;
+    }
+
+    // the chosen segments all need to be stable
+    const auto unstable = std::any_of(
+      range.first, range.second, [](ss::lw_shared_ptr<segment>& seg) {
+          return seg->has_appender();
+      });
+    if (unstable) {
+        return std::nullopt;
+    }
+
+    return range;
+}
+
+ss::future<> disk_log_impl::compact_adjacent_segments(
+  std::pair<segment_set::iterator, segment_set::iterator> range,
+  storage::compaction_config cfg) {
+    // lightweight copy of segments in range
+    std::vector<ss::lw_shared_ptr<segment>> segments;
+    std::copy(range.first, range.second, std::back_inserter(segments));
+
+    auto target = segments.front();
+
+    /*
+     * remove index files. they will be rebuilt by the single segment compaction
+     * operation, and also ensures we examine segments correctly on recovery.
+     */
+    if (co_await ss::file_exists(target->index().filename())) {
```

Comment:
Can compaction fail? Should these be moved to after or just before when mutations to the index occur?

Reply:
It can fail. Removing these indexes is a way to signal that they should be rebuilt.